woocommerce/woocommerce-ios
# Ensure IPP objects are deallocated after failure and cancellation code paths with unit tests

Closes: #6612
Part of #5983 (unit tests for `CollectOrderPaymentUseCase` and `RefundSubmissionUseCase`)


### Description


First, an apology for going over 500 diffs in this PR 🙇🏻‍♀️ I could have separated the changes for the refund case to another PR, but thought it might be more clear to see the same changes to both payment and refund use cases.

#### The failure code path

There are several paths to reach the end of the IPP flow, encapsulated by `CollectOrderPaymentUseCase`. Since `CollectOrderPaymentUseCase` holds on to `CardReaderConnectionController` which subscribes to the reader state, we want to make sure all these IPP objects are deallocated after the end of the payment flow. Otherwise, any new updates on the card reader (like connecting to the reader after switching to a new store) could trigger the previous card reader subscription and result in weird behaviors like described in #6612. This is currently done at the completion block of `CollectOrderPaymentUseCase.collectPayment`:

https://github.com/woocommerce/woocommerce-ios/blob/667d162af29c682c8e383441b4bbeec0debece1d/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsViewModel.swift#L542-L545

While adding unit tests for `CollectOrderPaymentUseCase` and `RefundSubmissionUseCase` in this PR, I noticed that the completion block for `CollectOrderPaymentUseCase.collectPayment` isn't invoked on the error case after dismissing an error alert. This is fixed by delegating the dismissal of the error alert from `CardPresentModalError` view model to the use case so that not only the alert UI is dismissed but also the completion block is called.

In addition to the completion block being invoked, there were retain cycles that caused the IPP objects to linger after the end of IPP flow. I took a pass to make sure to weak `self` in closures and the IPP objects are deallocated now.

#### The cancellation code path

Other than the failure scenario, I also noticed the completion block not being invoked after canceling the payment. This is now in a test case, and I also manually tested to make sure IPP objects are deallocated afterward using Xcode memory debugger.

#### Making payment/refund use case testable

In order to test the success/failure/cancellation scenarios of IPP, we need a way to interact with the alerts UI in unit tests. I added a protocol `OrderDetailsPaymentAlertsProtocol` so that we can pass the mock version to trigger different UI states. Please note that there are a lot more test cases for `CollectOrderPaymentUseCase` than `RefundSubmissionUseCase` due to the size of this PR. I will add more test cases when I implement analytics for the refund flow as part of https://github.com/woocommerce/woocommerce-ios/issues/5984.

### Testing instructions


Prerequisite: having two stores eligible for IPP
1. Switch to a site with an empty store address (you can update the store address in wp-admin > WooCommerce > Settings)
1. Go to the Orders tab and open an unpaid order
1. Tap on collect payment
1. Notice "Missing store address" alert is shown
1. Tap "Cancel" to dismiss the alert (before this PR, the completion block for this IPP flow isn't invoked due to how the error alert `CardPresentModalError` secondary action is set)
1. Switch to a store with working IPP
1. Enter IPP settings (Menu tab > Settings > In-Person Payments)
1. Connect to a reader --> the reader should be connected
1. Disconnect from a reader
1. Connect to a reader again --> the reader should be connected again
1. Go to the Orders tab and open an unpaid order
1. Collect payment --> the payment should go through

---
